src/averageAges.js

```diff
@@ -1,61 +1,51 @@
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
- * @param {object[]} people
- * @param {number} century - optional
- *
- * @return {number}
- */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const result = people.filter(person => person.sex === 'm'
+    && (!century || Math.ceil(person.died / 100) === century));
+  const finalResult = result.reduce((sum, person) =>
+    sum + person.died - person.born, 0);
+
+  return finalResult / result.length;
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const result = people.filter(person => person.sex === 'f'
+    && (!withChildren || people.find(children =>
+      children.mother === person.name)));
+  const finalResult = result.reduce((sum, person) =>
+    sum + person.died - person.born, 0);
+
+  return finalResult / result.length;
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothersObject = people.filter(person =>
+    person.sex === 'f').reduce((mothersAccumulated, mother) => {
+    return {
+      ...mothersAccumulated,
+      [mother.name]: {
+        born: mother.born,
+        childrenBornYears: people.filter(person =>
+          person.mother === mother.name).filter(person =>
+          onlyWithSon ? person.sex === 'm' : true).map(({ born }) => born),
+      },
+    };
+  }, {});
+
+  const motherNames = Object.keys(mothersObject);
+  let mothersTotal = 0;
+  const ageDiffs = motherNames.flatMap(motherName => {
+    const mother = mothersObject[motherName];
+
+    return mother.childrenBornYears.map(year => {
+      mothersTotal += 1;
+
+      return year - mother.born;
+    });
+  });
+
+  return ageDiffs.reduce((ages, age) => ages + age, 0) / mothersTotal;
```

Comment:
interesting solution! But consider changing the approach to start with finding all children's -> mother of each child -> calculate the difference. It should reduce calculations